### ISSUE TITLE
Problem when running the saigup example in opm-core

### ISSUE THREAD

**Issue by dlmorenob** [OPM/opm-core#841]
Good day,

I am trying to use the latest version of opm-core and to run the SAIGUP example provided by the MRST people some years ago.  The `sim_2p_incomp` simulator runs fine on the 2013.03 version, however, when I am running the same DATAFILE using the latest version of the library (current master) I get a segmentation fault (core dump), specifically when it reaches the instruction:
```C++
wellreport.push(props_, *wells_, state.saturation(), 0.0, well_state.bhp(), well_state.perfRates());
```
in the `SimulatorIncompTwophase.cpp` function.  Any idea why is this happening?

![screenshot from 2015-08-08 23 35 06](https://cloud.githubusercontent.com/assets/9252709/9153580/28c5024c-3e26-11e5-9600-28017e03aa9b.png)

![screenshot from 2015-08-08 23 31 55](https://cloud.githubusercontent.com/assets/9252709/9153573/b32df322-3e25-11e5-9549-4daf4e927862.png)

Is there something I am overlooking?  I appreciate the help.

**Comment by atgeirr** [comment]
I am not sure what is wrong. Could you send me the parameter file you used (to atgeirr@sintef.no) to run the case? What was the name of the deck used?

**Comment by dlmorenob** [comment]
Sure.  Thanks.  I am actually trying to run the Brugge case for I am trying
to use it for research purposes and I got the same (core dump) error.  I
decided then to try the Saigup example that used to work on 2013.03 but it
crashes as well on the latest release.  Could it be because of any library
that is outdated?   I attach the requested files. I appreciate any hints
or  help.

Thank you for the time.




2015-08-12 3:45 GMT-05:00 Atgeirr Flø Rasmussen <notifications@github.com>:

> I am not sure what is wrong. Could you send me the parameter file you used
> (to atgeirr@sintef.no) to run the case? What was the name of the deck
> used?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/OPM/opm-core/issues/841#issuecomment-130219998>.
>



-- 

*David Leonardo Moreno Bedoya*

**Comment by atgeirr** [comment]
I have found two bugs that caused segfaults, both relating to the recent introduction of the ExplicitArraysFluidState class:
 - In one else-branch setIndex() was not called.
 - It is hardcoded for three phases, but used also for two-phase cases.

I will try to fix this, and report here when it has been resolved. Note that in testing, after adding workarounds for the bugs I still could not complete the first step of the case you attached (it cannot satisfy the well constraints after 10 tries), but that is a different issue.

@andlaus I will propose a fix today.

**Comment by dlmorenob** [comment]
I see, thank you. I'll be attentive to the changes.  Maybe that might be why the Brugge case it is not running as well on this simulator.  Anyway.  I wanted to point out that I have just run the same case on the 2013.03 version and as happened to you it has some problems fulfilling the conditions of the wells.  However, after some tries it runs correctly.  Shouldn't both versions behave the same?

I attach a screen.

![screenshot from 2015-08-12 06 22 39](https://cloud.githubusercontent.com/assets/9252709/9222853/cdfc6160-40ba-11e5-9648-321f32673088.png)

One more thing.  I would like to be able to use a GUI for the compiling and following up of the code.  Do you recommend any (Eclipse?, QT?).  Thanks.

**Comment by atgeirr** [comment]
About GUI: some use Eclipse, I use just Emacs and Terminal myself.

This is the bottom part of my run: a group control (with target zero?) fails.
```
Iteration         Residual        Change in p
        0         0.0062425                 *
        1       2.39367e-08       7.65897e+07
        2       7.31783e-13           48026.3
        3       2.64654e-16           8.87994
        4                 *         0.0016006
Solved pressure in 4 iterations.
Pressure solver took:  2.81022 seconds.
Checking well conditions.
BHP limit violated for well I5:
BHP limit = 1.8e+08
BHP       = 3.90986e+08
Well controls not passed, solving again.

Iteration         Residual        Change in p
        0         0.0167813                 *
        1       4.46787e-07       2.10986e+08
        2       8.00873e-12            102321
        3       3.45047e-16           5.73994
        4                 *       0.000379403
Solved pressure in 4 iterations.
Pressure solver took:  2.80868 seconds.
Checking well conditions.
Group1 target not met for group P-GROUP
target = 0
rate   = -0.0170426
Well controls not passed, solving again.

Iteration         Residual        Change in p
        0       3.12382e-05                 *
        1       5.14739e-09       1.53374e+07
        2       2.61002e-13           7278.93
        3                 *          0.502096
Solved pressure in 3 iterations.
Pressure solver took:  2.2499 seconds.
Checking well conditions.
Group1 target not met for group P-GROUP
target = 0
rate   = -0.0172935
Well controls not passed, solving again.

Iteration         Residual        Change in p
        0       2.96961e-06                 *
        1       7.87663e-10       5.82127e+06
        2       4.31248e-14            1586.7
        3                 *          0.146498
Solved pressure in 3 iterations.
Pressure solver took:  2.26326 seconds.
Checking well conditions.
Group1 target not met for group P-GROUP
target = 0
rate   = -0.0174924
Well controls not passed, solving again.

Iteration         Residual        Change in p
        0       1.13261e-06                 *
        1        2.0009e-10       2.97523e+06
        2       1.12653e-14           657.475
        3                 *         0.0474729
Solved pressure in 3 iterations.
Pressure solver took:  2.26017 seconds.
Checking well conditions.
Group1 target not met for group P-GROUP
target = 0
rate   = -0.0176752
Well controls not passed, solving again.

Iteration         Residual        Change in p
        0       5.40247e-07                 *
        1       7.04936e-11       2.56319e+06
        2       3.89606e-15           504.404
        3                 *          0.016251
Solved pressure in 3 iterations.
Pressure solver took:  2.2816 seconds.
Checking well conditions.
Group1 target not met for group P-GROUP
target = 0
rate   = -0.017853
Well controls not passed, solving again.

Iteration         Residual        Change in p
        0       3.76977e-07                 *
        1       5.23142e-11       2.29965e+06
        2       1.56259e-15           414.821
        3                 *         0.0126486
Solved pressure in 3 iterations.
Pressure solver took:  2.24169 seconds.
Checking well conditions.
Group1 target not met for group P-GROUP
target = 0
rate   = -0.0180286
Program threw an exception: [/Users/atgeirr/opm-build-release/opm-core/opm/core/simulator/SimulatorIncompTwophase.cpp:535] Could not satisfy well conditions in 10 tries.
libc++abi.dylib: terminating with uncaught exception of type std::runtime_error: [/Users/atgeirr/opm-build-release/opm-core/opm/core/simulator/SimulatorIncompTwophase.cpp:535] Could not satisfy well conditions in 10 tries.
Abort trap: 6
```

**Comment by dlmorenob** [comment]
There are no group control specified on the datafile, the simulator is assuming 0 for the target of the groups.  Are they necessary? In Eclipse (slb) they are not..

**Comment by atgeirr** [comment]
The simulator is wrong to assume the zero group target (should be no target at all). This bug is handled in opm-parser, a PR will appear there to fix it. As a workaround you can temporarily change the file from opm-parser at `opm/parser/eclipse/EclipseState/Schedule/Group.cpp` lines 43-46 to have a negative target (-1e100 for example) rather than 0.

You must also update opm-core (that fix has been merged to the repo).

Please tell us if that makes everything work as before.

**Comment by joakim-hove** [comment]
> This bug is handled in opm-parser, a PR will appear there to fix it

OK - this was 'fixed' in: https://github.com/OPM/opm-parser/pull/553.

**Comment by dlmorenob** [comment]
There is a missing file named  --- disable_warnings.h ---- that is supposed to be on the platform_dependent folder (the folder is not on the repo).
![screenshot from 2015-08-12 08 42 40](https://cloud.githubusercontent.com/assets/9252709/9225302/45ce7904-40ce-11e5-984d-90d0386f7106.png)

![screenshot from 2015-08-12 08 44 29](https://cloud.githubusercontent.com/assets/9252709/9225327/5f1979cc-40ce-11e5-9452-b239042ce45b.png)

**Comment by joakim-hove** [comment]
Fetch latest and gratest opm-parser and renhold that first; the file in
question has moved from opm-core to opm-parser.
12. aug. 2015 15:44 skrev "David Leonardo Moreno Bedoya" <
notifications@github.com> følgende:

**Comment by alfbr** [comment]
On occasions I stray away from emacs and cli in favour of a GUI. My favourite for a project like opm is kdevelop. It has nice integration of c++, cmake, gdb and git.

**Comment by dlmorenob** [comment]
Thank you for the help.  The sim_2p_incomp is running now.  However, I still get the same error and it seems the fix for the groups did not work. 

This is the screen of the newly compile code.
![screenshot from 2015-08-12 09 58 01](https://cloud.githubusercontent.com/assets/9252709/9227408/fe486fd0-40d8-11e5-9f01-d930802d53d3.png)

and this one is the one from 2013.03

![screenshot from 2015-08-12 09 59 20](https://cloud.githubusercontent.com/assets/9252709/9227417/085f402a-40d9-11e5-8fe4-49a7460c7468.png)

I checked the opm-parser - Group.cpp file and I see the change suggested was implemented but there is still something wrong.

![screenshot from 2015-08-12 10 01 41](https://cloud.githubusercontent.com/assets/9252709/9227499/73e9142e-40d9-11e5-82cb-fe8bbc4b40d7.png)

Thanks.

PS.  Thank you for the tip.  I'll check the GUI.

**Comment by atgeirr** [comment]
> I checked the opm-parser - Group.cpp file and I see the change suggested was implemented but there is still something wrong.

Probably I was too quick when looking at this, will try to figure out a proper cure tomorrow.

**Comment by dlmorenob** [comment]
I appreciate.  I will try to follow the error  as well and to show you some more when running the Brugge case. They are different.  Also with sim_2p_incomp.

**Comment by atgeirr** [comment]
You could try checking out the branches referred to in OPM/opm-parser#554 and OPM/opm-core#850 (they have not been merged yet).

**Comment by atgeirr** [comment]
Now that the fixes has been merged to the master branches the simulator should work as before for you. Can you verify this @dlmorenob?

**Comment by dlmorenob** [comment]
I will check.

Thanks.

2015-08-14 7:06 GMT-05:00 Atgeirr Flø Rasmussen <notifications@github.com>:

> Now that the fixes has been merged to the master branches the simulator
> should work as before for you. Can you verify this @dlmorenob
> <https://github.com/dlmorenob>?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/OPM/opm-core/issues/841#issuecomment-131087300>.
>



-- 

*David Leonardo Moreno Bedoya*

**Comment by dlmorenob** [comment]
It is working fine now.

![screenshot from 2015-08-14 09 31 30](https://cloud.githubusercontent.com/assets/9252709/9276057/51ff5168-4267-11e5-971f-559aacc18b6d.png)

I have one question though, if I may.  What is the purpose of >> Starting the report step # >> in the end?  I do not see any additional files being created, am I wrong?  also, I see a small error there, the number of time steps is always 0.  

Thanks.

**Comment by atgeirr** [comment]
> I have one question though, if I may. What is the purpose of >> Starting the report step # >> in the end? I do not see any additional files being created, am I wrong? also, I see a small error there, the number of time steps is always 0.

That is a minor bug as well, I'll see if I can fix it.

> Also, I see that the Wellreport.txt contains much less information than before. Is there a way to control this not within the code? perhaps the Max_well_control_iterations parameter?

No, that parameter controls how many times to allow changing well control before giving up, when trying to satisfy the given constraints. I'll look into the wellreport.txt issue next week.

**Comment by dlmorenob** [comment]
Thanks

**Comment by dlmorenob** [comment]
I have been trying to run a case called the Brugge case on the sim_2p_incomp simulator since it is a two phase problem and with incompressible properties as well.  Basically, I am trying to reproduce some work related to SPE paper No. 119094.  Anyway.  I tried to run the case on the 2013.03 version of the simulator and I get the following error.

![screenshot from 2015-08-14 11 27 37](https://cloud.githubusercontent.com/assets/9252709/9278623/8ea11556-4277-11e5-9930-f22b2b3c5fae.png)

When I try to run the same datafile on the current version I get some even nastier failures. 

![screenshot from 2015-08-14 11 30 44](https://cloud.githubusercontent.com/assets/9252709/9278684/0df0d364-4278-11e5-86d7-650982574a6d.png)

and this one

![screenshot from 2015-08-14 11 30 57](https://cloud.githubusercontent.com/assets/9252709/9278683/0ddb9760-4278-11e5-89cf-526ba204c4f1.png)

Any ideas? I have started with opm-core but would you suggest to try opm-autodiff instead meanwhile?  I would like to help debugging all this but I will take some time to get familiar with the code and the system.  Meantime, I appreciate any help that might be provided since we are planning to use the simulator for performing the simulations of the Brugge model for some  academic work.

Thank you for the time.

**Comment by atgeirr** [comment]
The "flow" program in opm-core is at this point a quite robust simulator. It uses a fully-implicit approach, whereas the older two-phase sims use an operator-splitting sequential approach (implicit pressure, reordered implicit transport), so "flow" may be slower in cases that are well-suited for the splitting solvers. Currently the two-phase support in "flow" is broken, but if you have a three-phase case you could try flow. It is possible that extending a case to three phases by adding the GAS keyword and the necessary pvt and saturation functions (PVDG and SGOF) to the deck.

The transport part of the operator split sim is designed to be very robust, and the errors you are seeing is probably a bug. If you are free to share the data file I can take a look at the error.

**Comment by dlmorenob** [comment]
Thank you.  I indeed knew the flow program was a 3 phase sim. and I thought sim_2p was good enough.  I think once it works it will be enough for now.  Although, I could also try augmenting to three phases as you suggest and using the implicit one. 

Thanks for the help.

**Comment by alfbr** [comment]
We are in the process of setting up SPE10 (which is two two-phase models) to run with flow. We will share the full decks under the open database license once done. Pop me a mail, and we can give you the current deck. It should help you fill in the missing pieces to run Brugge with flow.

**Comment by atgeirr** [comment]
The saigup example (and 2-phase in general) broke again in September (during further reorganization of the saturation functions), but has now been fixed in #910. Closing this issue. Feel free to open new issues about related things!
